src/utils/TrustedForwarder.sol

```diff
@@ -0,0 +1,120 @@
+            if (!_isTrustedByTarget(calli.target)) {
+                revert TargetNotTrusted(calli.target);
+            }
+
+            // define the allocated value for this call
+            uint256 val = calli.value;
+
+            // add the allocated value to the value accumulator
+            // so we can check the msg.value at the end
+            // to ensure it matches the sum of the call values
+            valAccumulator += val;
+
+            // make the call and record the result data and whether it was successful
+            (bool success, bytes memory data) = calli.target.call{value: val}(
+                abi.encodePacked(calli.callData, msg.sender)
+            );
+
+            // *ALWAYS* revert if the call was not successful
+            /// @dev prevents ETH from being locked in the contract if call
+            /// with non-zero value reverts
+            if (!success) {
+                bytes memory revertData = data;
+                uint256 len = data.length;
+                assembly {
+                    revert(add(revertData, 0x20), len)
+                }
+            }
+
+            returnData[i] = data;
+
+            unchecked {
+                ++i;
+            }
+        }
+
+        // Post iteration, make sure the msg.value == SUM(call[0...i].value)
+        if (msg.value != valAccumulator) {
+            revert MismatchedValue(valAccumulator, msg.value);
+        }
+    }
+
+    /// @dev Returns whether the target trusts this forwarder.
+    /// This function performs a static call to the target contract calling the
+    /// {ERC2771Context-isTrustedForwarder} function.
+    function _isTrustedByTarget(address target) internal view returns (bool) {
+        bytes memory encodedParams =
+            abi.encodeCall(ERC2771Context.isTrustedForwarder, (address(this)));
+
+        bool success;
+        uint256 returnSize;
+        uint256 returnValue;
+        /// @solidity memory-safe-assembly
+        assembly {
+            // Perform the staticcal and save the result in the scratch space.
+            // | Location  | Content  | Content (Hex)                                                      |
+            // |-----------|----------|--------------------------------------------------------------------|
+            // |           |          |                                                           result ↓ |
+            // | 0x00:0x1F | selector | 0x0000000000000000000000000000000000000000000000000000000000000001 |
+            success :=
+                staticcall(
+                    gas(),
+                    target,
+                    add(encodedParams, 0x20),
+                    mload(encodedParams),
+                    0,
+                    0x20
+                )
+            returnSize := returndatasize()
+            returnValue := mload(0)
+        }
+
+        return success && returnSize >= 0x20 && returnValue > 0;
```

Comment:
**Question:** Should we allow a situation where `returnSize > 0x20` and `returnValue != 1`? It seems to me that `returnSize == 0x20 && returnValue == 1` should always be true.